### DISCUSSION TITLE
Streaming Status Code: 103 Early Hints

### DISCUSSION THREAD

**Discussion by mangs** [solidjs/solid#1026] [Ideas]
### Idea

Hi all,
I've been doing a lot of thinking about streaming and performance, and I have some ideas about using streaming in potentially different ways using the HTTP `103 Early Hints` status code.

A big reason to use streaming is to get content sent to the browser as soon as possible, usually before a page has finished fully rendering. And the biggest benefit is usually rendering the `<head>` as soon as possible so dependent resources are downloaded quickly. However, doing so comes by sacrificing the ability to determine a proper status code after building the page contents; rather, a `200 OK` response must be returned while hoping that an error doesn't occur while streaming the response.

A `200`+ HTTP response contains a 1:1 mapping of status code to response, whereas [`1xx` responses are "informational"](https://www.rfc-editor.org/rfc/rfc7231.html#section-6.2) meaning they are only a subset or for argument purposes a "chunk" of the full response; the `103 Early Hints` response builds on top of the latter concept. If we look at [the spec for 103 Early Hints](https://httpwg.org/specs/rfc8297.html), notice the output of the example response:

```
HTTP/1.1 103 Early Hints
Link: </main.css>; rel=preload; as=style

HTTP/1.1 103 Early Hints
Link: </style.css>; rel=preload; as=style
Link: </script.js>; rel=preload; as=script

HTTP/1.1 200 OK
Date: Fri, 26 May 2017 10:02:11 GMT
Content-Length: 1234
Content-Type: text/html; charset=utf-8
Link: </main.css>; rel=preload; as=style
Link: </newstyle.css>; rel=preload; as=style
Link: </script.js>; rel=preload; as=script

<!doctype html>
[... rest of the response body is omitted from the example ...]
```

Similar to a streaming response, each block of status code plus headers is delivered in "chunks" until the final `200`+ status code arrives which signifies the completion of the response. In the case where the primary use for streaming is to deliver assets from `<head>` as soon as possible, adding support for `103 Early Hints` will allow for all the current benefits of streaming rendering while allowing the application to decide as late as possible what response code to return.

### Server and Browser Support
Chrome is adding support for `103 Early Hints` in [version 103 coincidentally enough](https://blog.chromium.org/2022/05/chrome-103-beta-early-navigation-hints.html), and Cloudflare is supporting it in [beta now](https://blog.cloudflare.com/early-hints/).

Here are the bug reports tracking implementations in Chromium, Firefox, and WebKit:
- https://bugs.chromium.org/p/chromium/issues/detail?id=671310
- https://bugzilla.mozilla.org/show_bug.cgi?id=1407355
- https://bugs.webkit.org/show_bug.cgi?id=238921

### Should We Support This?
I really think so. If the team agrees, I'd love for this to be my first contribution to the project. 👌

**Comment by ryansolid** [comment]
This feels like something that would be implemented metaframework-wise since that's where asset awareness is (like SolidStart). I actually mocked it up a couple of months ago because I heard of the Cloudflare support, but I couldn't find the "speed tab" they are referring to. I went ahead and tested a bunch of stuff but it never worked so I just left it alone. It's pretty easy to implement on top of Solid's `renderToStringAsync`.

In any case, excited this is coming to Chrome soon.

**Comment by mangs** [comment]
Nice. Can I help add it to SolidStart if what you did isn't already in working shape?

As for where it is in the Cloudflare dashboard, you can see it by following [this link](https://dash.cloudflare.com/?to=/:account/:zone/speed/optimization#early_hints) then scrolling down to the Early Hints box to enable the beta for your account.
